### PR TITLE
Set ChainId and GasePrice in storage

### PR DESCRIPTION
**Pull request by tgmichel**
- Define `ChainId` and `GasPrice` storage values.
- Set an optional genesis config or fallback to default (ChainId `42` and GasPrice `zero`).

### CODE REVIEW

**Comment by xlc** [comment]
I can see why you want to configure this is chain spec but this is going to add performance penalty for extra storage reads on every transactions

File: frame/evm/src/lib.rs [hunk 1]
```diff
@@ -231,14 +229,26 @@ pub struct GenesisAccount {
...
+		config(gas_price): Option<U256>;
 		build(|config: &GenesisConfig| {
+			match &config.chain_id {
+				Some(chain_id) => ChainId::put(chain_id),
+				_ => ChainId::put(42)
+			};
+			match &config.gas_price {
+				Some(gas_price) => GasPrice::put(gas_price),
+				_ => GasPrice::put(U256::zero())
+			};
```
Comment (bkchr) [inline frame/evm/src/lib.rs:251]:
```suggestion
```

File: frame/evm/src/lib.rs [hunk 2]
```diff
@@ -231,14 +229,26 @@ pub struct GenesisAccount {
...
+		GasPrice get(fn gas_price): U256;
 		AccountCodes get(fn account_codes): map hasher(blake2_128_concat) H160 => Vec<u8>;
 		AccountStorages get(fn account_storages):
 			double_map hasher(blake2_128_concat) H160, hasher(blake2_128_concat) H256 => H256;
 	}
 
 	add_extra_genesis {
 		config(accounts): std::collections::BTreeMap<H160, GenesisAccount>;
+		config(chain_id): Option<u64>;
+		config(gas_price): Option<U256>;
```
Comment (bkchr) [inline frame/evm/src/lib.rs:242]:
```suggestion
```

File: frame/evm/src/lib.rs [hunk 3]
```diff
@@ -231,14 +229,26 @@ pub struct GenesisAccount {
 
 decl_storage! {
 	trait Store for Module<T: Trait> as EVM {
+		ChainId get(fn chain_id): u64;
+		GasPrice get(fn gas_price): U256;
```
Comment (bkchr) [inline frame/evm/src/lib.rs:233]:
```suggestion
		ChainId get(fn chain_id) config(): u64 = 42;
		GasPrice get(fn gas_price) config(): U256;
```

**Review by bkchr — COMMENTED** [review]
This is doing the same as you have done there manually.

**Comment by tgmichel** [comment]
> I can see why you want to configure this is chain spec but this is going to add performance penalty for extra storage reads on every transactions

As far as I understand runtime accessing it's state for a StorageValue is very cheap, do you think this will add any _noticeable_ overhead when processing a transaction?

Anyway, this change is directly related to the frontier PR you are aware of, and in that case the discussion is more on runtime spawn overhead VS. statebackend query on RPC calls that are unlimited and totally free. This change will open a way of making those requests much more cheaper.

File: frame/evm/src/lib.rs (hunk 3 above)
Comment (tgmichel) [reply frame/evm/src/lib.rs]: Thanks, you don't know macro magic until you do!

**Comment by xlc** [comment]
Both of chain ID and min gas are constants that I don't feel it make sense to make them into storage.
Shouldn't the Frontier RPC handler able to cache the value to avoid the overhead?

**Comment by tgmichel** [comment]
> Both of chain ID and min gas are constants that I don't feel it make sense to make them into storage.

Again, all this makes sense in the context of the https://github.com/paritytech/frontier/pull/175. 

> Shouldn't the Frontier RPC handler able to cache the value to avoid the overhead?

Although I talked with @sorpaas about the above Frontier PR and this one before opening them and he seemed OK with the changes, I'm open to switch to a more flavoured option like the one you suggested there - enabling both runtime and storage APIs - if he thinks that's more reasonable. If that's the case, a cache strategy for ChainId and Gasprice in the RPC side may be preferable, as it'd support both API _flavours_ without the overhead of accessing the StorageValue.

**Comment by xlc** [comment]
In memory cache should always give your better performance (except for the first call) if that's the purpose, especially for constant data.

But I am not too against on this as we are considering on fork pallet-evm anyway to experiment some ideas like state renting.

**Comment by JoshOrndorff** [comment]
Before this PR, pallet evm used the `Get` trait to obtain a chainId. That design is strictly more flexible, and in my opinion should be kept. It allows the simplicity and performance that @xlc described by simply using the `parameter_types!` macro. It also allows the flexibility of storing the chainId in a storage item, we just need to introduce a simple auxiliary pallet that stores the chainId.

IMO putting the storage item into pallet-evm forces _everyone_ to make the same decision, while using the `Get` trait allows the decision to be configured.

**Comment by tgmichel** [comment]
Closing this. Rel https://github.com/paritytech/frontier/pull/199

**Pull request closed without merging**